src/rgw/rgw_http_client_curl.cc

```diff
@@ -0,0 +1,103 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+extern "C" {
+#include <curl/curl.h>
+}
+
+#include "rgw_http_client_curl.h"
+#include <mutex>
+#include <vector>
+
+#include "rgw_common.h"
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rgw
+
+#ifdef WITH_RADOSGW_OPENSSL_LOCK_COMPAT
+#include <openssl/crypto.h>
+#endif
+
+namespace rgw {
```

Comment:
Since we're now using C++17, you can also write:
namespace rgw::curl {

Reply:
My hope is to backport some of this to Luminous and Jewel since the bug affects all versions of rgw atm